README

= theruck


== Description


== Installation

=== Archive Installation

	rake install

=== Gem Installation

	gem install theruck


== Features/Problems


== Synopsis


== Copyright

Author::    cho45 <cho45@lowreal.net>
Copyright:: Copyright (c) 2008 cho45
License::   


ChangeLog
== 0.0.1 / 2008-05-25

* initial release



spec/session_spec.rb
#!/usr/bin/env ruby

require "pathname"

$LOAD_PATH << Pathname.new(__FILE__).parent + "fixtures"
$LOAD_PATH << Pathname.new(__FILE__).parent.parent + "lib"

require "rubygems"
require "spec"
require "theruck"
include TheRuck

class TestRootSessionController < Controller
	include TheRuck::Session

	class AdminController < Controller
		before do
			redirect "/login" unless session["user"]
		end

		route "sample1" do
			body "sample1"
		end

		after do
			body "after"
		end
	end

	route "" do
		head "Content-Type", "text/plain"
		body "index"
	end

	route "login" do
		session["user"] = {
			"name" => "foo"
		}
	end

	route "admin/*" => :AdminController
end

def warn(*)
	# pass
end

describe TheRuck do
	before do
		@req = Rack::MockRequest.new(TestRootSessionController)
	end

	it "should have session" do
		res =  @req.get("/admin/sample1")
		res.status.should == 302
		res.headers["Location"].should == "/login"
		cookie = res.headers["Set-Cookie"]

		res =  @req.get("/login", "HTTP_COOKIE" => cookie)
		res.status.should == 200

		res =  @req.get("/admin/sample1", "HTTP_COOKIE" => cookie)
		res.status.should == 200
	end
end



spec/theruck_spec.rb
#!/usr/bin/env ruby


$LOAD_PATH << "lib"
$LOAD_PATH << "../lib"

require "theruck"
include TheRuck

class HTML < View
	def result
	end
end

class Foo < Controller
	# view :HTML # default view

	route "" do
		head "Content-Type", "text/plain"
		body "index"
	end

	route "help" do
		head "Content-Type", "text/plain"
		body "help"
	end

	route "view" do
		stash.json = [ 1, 2, 3 ]
	#	view :foo, JSON
	end

	route "method", GET do
		stash.body = "get"
	end

	route "method", PUT do
		stash.body = "put"
	end

	route "method", POST do
		stash.body = "post"
	end

	route "foo/:param1/:param2" do
		stash.json = params.inject({}) {|r,(k,v)| r.update(k.to_s => v) }
	#	view :foo, JSON
	end

	route "api" => :ApiController
end

# view and controller has same interface.
describe TheRuck do
	it "view" do
	end

	it "controller" do
		Rack::MockRequest
		Foo.new(env).result
	end
end



spec/controller_spec.rb
#!/usr/bin/env ruby

require "pathname"

$LOAD_PATH << Pathname.new(__FILE__).parent + "fixtures"
$LOAD_PATH << Pathname.new(__FILE__).parent.parent + "lib"

require "rubygems"
require "spec"
require "theruck"
include TheRuck

class TestRootController < Controller
	route "" do
		head "Content-Type", "text/plain"
		body "index #{normal_method}"
	end

	route "help" do
		head "Content-Type", "text/plain"
		body "help"
	end

	route "detach" do
		detach "/"
		body "not shown"
	end

	route "method", GET do
		head "Content-Type", "text/plain"
		body "get"
	end

	route "method", PUT do
		head "Content-Type", "text/plain"
		body "put"
	end

	route "method", POST do
		head "Content-Type", "text/plain"
		body "post"
	end

	route "params/:param1/:param2" do
		head "Content-Type", "application/octet-stream"
		body Marshal.dump(params)
	end

	route "regexp/:year #{/\d{4}/}" do
		head "Content-Type", "application/octet-stream"
		body Marshal.dump(params)
	end

	route "regexp/:year #{/\d{4}/}/:month #{/\d\d/}" do
		head "Content-Type", "application/octet-stream"
		body Marshal.dump(params)
	end

	route "regexp/:year #{/\d{4}/}/:month #{/\d\d/}/:day #{/\d\d/}" do
		head "Content-Type", "application/octet-stream"
		body Marshal.dump(params)
	end

	route "double-match/*" do
		body "double-match1"
	end

	route "double-match/foo/*" do
		body "double-match2"
	end

	route "api/*"        => :ApiController
	route "api1/:user/*" => :ApiController

	def normal_method
		"foobar"
	end
end

def warn(*)
	# pass
end

describe TheRuck do
	before do
		@req = Rack::MockRequest.new(TestRootController)
	end

	it "should dispatch index and response correctly" do
		res = @req.get("")
		res.headers["Content-Type"].should == "text/plain"
		res.body.should  == "index foobar"
	end

	it "should treat slashes correctly" do
		@req.get("/help").body.should == @req.get("/help/").body
	end

	it "should dispatch each http method." do
		@req.get("/method").body.should == "get"
		@req.put("/method").body.should == "put"
		@req.post("/method").body.should == "post"
	end

	it "should have ordered match" do
		@req.get("/double-match").body.should == "double-match1"
		@req.get("/double-match/foo").body.should == "double-match2"
		@req.get("/double-match/bar").body.should == "double-match1"
	end


	it "should pass parameters via params method" do
		Marshal.load(@req.get("/params/foo/bar").body).should == {"param1"=>"foo", "param2"=>"bar"}
		Marshal.load(@req.get("/params/foo/bar?opt=1").body).should == {"param1"=>"foo", "param2"=>"bar", "opt"=>"1"}
	end

	it "should handle regexp param" do
		@req.get("/regexp/xxxx").status.should == 404
		Marshal.load(@req.get("/regexp/2005").body).should == {"year"=>"2005"}
		Marshal.load(@req.get("/regexp/2005/01").body).should == {"year"=>"2005", "month"=>"01"}
		Marshal.load(@req.get("/regexp/2005/01/21").body).should == {"year"=>"2005", "month"=>"01", "day"=>"21"}
	end

	it "should autoload sub controllers" do
		TestRootController.autoload?(:ApiController).should == "api_controller"
	end

	it "should dispatch sub controllers correctly" do
		@req.get("/api/sample1").body.should == "sample1"
		Marshal.load(@req.get("/api/params/foo/bar").body).should == {"param1"=>"foo", "param2"=>"bar"}
		Marshal.load(@req.get("/api1/cho45/params/foo/bar").body).should == {"user"=>"cho45","param1"=>"foo", "param2"=>"bar"}
	end

	it "should handle detach action" do
		@req.get("/detach").body.should == "index foobar"
		@req.get("/api/detach").body.should == "index foobar"
	end

	it "should have stash object" do
		stash = Stash.new
		stash.json = [1, 2, 3]

		stash.should == { "json" => [1, 2, 3] }
		stash.json.should == stash["json"]
	end
end



examples/plusplus.rb
#!/opt/local/bin/ruby
#!/usr/bin/env ruby

require "rubygems"

$LOAD_PATH << "/Users/cho45/project/theruck/lib"
$LOAD_PATH << "gemlib"
require "theruck"
include TheRuck

require "gdbm"

class PlusPlusRoot < Controller
	@@name = "plusplus.db"

	route "" do
		res = nil
		db do |dbm|
			res = dbm.keys.map {|i| i.sub(/(\-\-|\+\+)$/, '') }.uniq.map {|nick|
				pp = dbm.fetch("#{nick}++", 0).to_i
				mm = dbm.fetch("#{nick}--", 0).to_i
				[nick, pp - mm, pp, mm]
			}.sort_by {|nick, karma, pp, mm| karma }
		end

		head "Content-Type", "text/plain"

		body "Top:\n"
		res.last(10).reverse_each do |nick, karma, pp, mm|
			break if karma < 0
			body "#{nick}: #{karma} (#{pp}++ #{mm}--)\n"
		end

		body "\n"
		body "Worst:\n"
		res.first(10).each do |nick, karma, pp, mm|
			break if karma > 0
			body "#{nick}: #{karma} (#{pp}++ #{mm}--)\n"
		end

		body "\n"
		body "Pop:\n"
		res.sort_by {|nick, karma, pp, mm| pp + mm}.last(10).reverse_each do |nick, karma, pp, mm|
			body "#{nick}: #{pp + mm} // #{karma} (#{pp}++ #{mm}--)\n"
		end

	end

	route ":nick" do
		nick = params["nick"]
		res = nil
		db do |dbm|
			pp = dbm.fetch("#{nick}++", 0).to_i
			mm = dbm.fetch("#{nick}--", 0).to_i
			res = [nick, pp - mm, pp, mm]
		end
		nick, karma, pp, mm = *res

		head "Content-Type", "text/plain"
		body "#{nick}: #{karma} (#{pp}++ #{mm}--)\n"
	end

	# /foo++, /foo--
	route ":nick #{/.+(?:\+\+|\-\-)$/}" do
		nick = params["nick"]
		nick.sub!(/(\-\-|\+\+)$/, '')
		redirect "/#{nick}"
	end

	# /foo++, /foo--
	route ":nick #{/.+(?:\+\+|\-\-)$/}", PUT do
		nick = params["nick"]

		res = nil
		db do |dbm|
			dbm[nick] = (dbm.fetch(nick, 0).to_i + 1).to_s
			nick.sub!(/(\-\-|\+\+)$/, '')
			pp = dbm.fetch("#{nick}++", 0).to_i
			mm = dbm.fetch("#{nick}--", 0).to_i
			res = [nick, pp - mm, pp, mm]
		end
		nick, karma, pp, mm = *res

		head "Content-Type", "text/plain"
		body "#{nick}: #{karma} (#{pp}++ #{mm}--)"
	end

	def db(&block)
		begin
			GDBM.open(@@name) do |dbm|
				yield dbm
			end
		rescue GDBMError, Errno::EWOULDBLOCK, Errno::EAGAIN
			retry
		end
	end
end

Plusplus = PlusPlusRoot # for rackup
if $0 == __FILE__
	Rack::Handler::CGI.run PlusPlusRoot, :Port => 4000
	# rackup -s webrick ./plusplus.rb
end


lib/theruck/session.rb
module TheRuck::Session # TODO
	def self.included(mod)
		mod.before do
			load_session
		end

		mod.after do
			save_session
		end
	end

	attr_accessor :session

	def load_session
		req = Rack::Request.new(env)
		@session_id = req.cookies["session_id"] || new_session_id
		@session = File.open(session_path) {|f| Marshal.load(f) } rescue {}
	end

	def save_session
		File.open(session_path, "w") {|f|  Marshal.dump(@session, f) }
		cookie = {
			:value => @session_id
		}
		# cookie[:expires] = Time.now + options[:expire_after] unless options[:expire_after].nil?
		res = Rack::Response.new(@body, @status, @header)
		res.set_cookie("session_id", cookie)
		@status, @header, @body = *res.to_a
	end

	def new_session_id
		@new_session = true
		require "digest/sha1"
		d = Digest::SHA1.new
		d.update(Time.now.to_s)
		d.update(Time.now.usec.to_s)
		d.update(rand(0).to_s)
		d.hexdigest
	end

	def session_path
		"/tmp/theruck_#{Digest::SHA1.hexdigest(@session_id)}"
	end
end


lib/theruck.rb

require "rack"

module TheRuck
	VERSION = "0.0.0"

	class TheRuckError < StandardError; end
	class Detach < Exception; end

	class Stash < Hash
		def method_missing(name, value=nil)
			name = name.to_s
			if name.sub!(/=$/, '')
				self[name] = value
			else
				self[name]
			end
		end
	end

	autoload :Session, "theruck/session"

	class Controller
		GET  = "GET"
		PUT  = "PUT"
		POST = "POST"
		HEAD = "HEAD"

		@@config = {}

		class << self
			attr_accessor :handlers, :before_handlers, :after_handlers

			def inherited(subclass)
				subclass.handlers        = []
				subclass.before_handlers = []
				subclass.after_handlers  = []
			end

			def config=(config)
				@@config = config
				self
			end

			def config
				@@config
			end

			def route(o, h=nil, &block)
				case o
				when String
					# route "foo"
					# route "foo", GET
					source = o
					method = h ? h : //

					handler_name = "handler_#{source}"
					handler_name << "_#{method}" if h
					define_method(handler_name, block)

					regexp, names = _route(source)
					self.handlers.unshift [source, regexp, method, names, handler_name]
				when Hash
					# route "api" => :ApiController
					source = o.keys.first
					classn, feature = o[source]
					feature = classn.to_s.scan(/[A-Z][^A-Z]*/).map {|i| i.downcase }.join("_") unless feature

					regexp, names = _route(source)
					self.handlers.unshift [source, regexp, //, names, classn]
					self.autoload classn, feature
				end
			end

			def _route(str)
				names = []
				paths = str.to_s.split("/", -1)
				regex = paths.empty?? %r|^/$| : Regexp.new(paths.inject("^") {|r,i|
					case i[0]
					when ?:
						name, regexp = i.sub(":", "").split(/\s+/, 2)
						names << name
						r << (regexp ? "/(#{regexp})" : "/([^/]+)")
					when ?*
						names << i.sub("*", "")
						r << "(?:/(.*))?"
					else
						r << "/#{i}"
					end
				} + "/?$")
				[regex, names]
			end

			def view(classn)

			end

			def before(&block)
				handler_name = "before_#{self.before_handlers.size}"
				define_method(handler_name, block)
				self.before_handlers << handler_name
			end

			def after(&block)
				handler_name = "after_#{self.after_handlers.size}"
				define_method(handler_name, block)
				self.after_handlers << handler_name
			end

			# Rack interface
			def call(env)
				new.handle(env)
			end
		end

		def initialize(parent=nil)
			@parent = parent
			if @parent
				@params = parent.params
			else
				@params = {}
			end
		end

		def handle(env)
			@status, @header, @body = 200, {}, []
			@stash  = Stash.new
			@env    = env
			@params.update env["QUERY_STRING"].split(/[&;]/).inject({}) {|r,pair|
				key, value = pair.split("=", 2).map {|str|
					str.tr("+", " ").gsub(/(?:%[0-9a-fA-F]{2})+/) {
						[Regexp.last_match[0].delete("%")].pack("H*")
					}
				}
				r.update(key => value)
			}

			self.class.before_handlers.each do |handler|
				send(handler)
			end

			dispatched = false
			self.class.handlers.each do |source, regexp, method, names, handler|
				# p [source, regexp, method, names, handler]
				if method === env["REQUEST_METHOD"] && regexp === env["PATH_INFO"]
					@params.update names.zip(Regexp.last_match.captures).inject({}) {|r,(k,v)|
						r.update(k => v)
					}
					dispatched = true
					case handler
					when Symbol
						warn "dispatch #{env["PATH_INFO"]} => #{source} => #{handler}"
						env["PATH_INFO"] = "/#{@params.delete("")}"
						@status, @header, body = self.class.const_get(handler).new(self).handle(env)
						@body.concat body
					else
						warn "dispatch #{env["PATH_INFO"]} => #{source} => #{handler}"
						send(handler)
					end
					break
				end
			end

			send("handler_default") unless dispatched

			self.class.after_handlers.each do |handler|
				send(handler)
			end

			[@status, @header, @body]
		rescue Detach => e
			if @parent
				raise
			else
				env["PATH_INFO"] = e.message
				retry
			end
		end

		def head(key, value=nil)
			case key
			when Hash
				@header.update key
			when Integer
				@status = key
			else
				@header[key] = value
			end
		end

		def body(l)
			@body << l
		end

		def params
			@params
		end

		def env
			@env
		end

		def handler_default
			head 404
			head "Content-Type", "text/plain"
			body "404"
		end

		def detach(path)
			raise Detach, path
		end

		def redirect(path)
			head 302
			# "#{env["rack.url_scheme"]}://#{env["SERVER_NAME"]}#{env["SCRIPT_NAME"]}#{path}"
			head "Content-Type", "text/html"
			head "Location", "#{env["SCRIPT_NAME"]}#{path}"
			body "<a href='#{env["SCRIPT_NAME"]}#{path}'>redirect</a>"
		end

		def method_missing(name, *args)
			if @parent
				@parent.send(name, *args)
			else
				super
			end
		end
	end

#	class View
#		class ErubisEruby < View
#			@@templates = {}
#
#			def initialize(opts={})
#				require "erubis"
#				super
#				@opts = {
#					:dir => "templates"
#				}.update(opts)
#				@layout = []
#				extend @opts[:helper] if @opts[:helper]
#			end
#
#			def render(_path, _stash)
#				@@templates[_path] ||= ::Erubis::EscapedEruby.new(File.read("#{@opts[:dir]}/#{_path}.html"))
#				head "Content-Type", "text/html"
#				_binding = binding
#				_stash.each {|k,v| eval "#{k} = _stash[:#{k}]", _binding }
#				body @layout.inject(@@templates[_path].result(binding)) {|content,layout|
#					@@templates[layout].result(binding)
#				}
#				self
#			end
#
#			def layout(path=:layout)
#				@@templates[path] ||= ::Erubis::EscapedEruby.new(File.read("#{@opts[:dir]}/#{path}.html"))
#				@layout << path
#			end
#		end
#
#		class JSON < View
#			def initialize(opts={})
#				require "json"
#				super
#			end
#
#			def render(path, stash)
#				head "Content-Type", "text/javascript"
#				body ::JSON.dump(stash)
#			end
#		end
#	end
end

